Add GSoC 2022 page (#34)
rizinorg/website: content/gsoc/2022.md

---
title: "GSoC 2022"
layout: "single"
ShowToc: true
TocOpen: false
ShowReadingTime: false

---

**TL;DR** Jump to the [Ideas list](#project-ideas).

# Introduction

This year is a second we participate as a fork - Rizin, effectively continuing the tradition since the 2015 year (as a radare2 project).

## Mentors

Members of the Rizin and Cutter core teams have volunteered to guide participants for GSoC’22. They were already guiding the participants for the GSoC and RSoC in past years. Please feel free to reach out to any of them in case you need any help in selecting a project.

- **Anton Kochkov** Mattermost: xvilka -- [@akochkov](https://twitter.com/akochkov)

> - **Riccardo Shirone** Mattermost: ret2libc

Reviewer: ```suggestion
- **Riccardo Schirone** Mattermost: ret2libc
```

# Project Ideas

## Rizin

## RZIL uplifting migration (Extended deadline)

Rizin has had the intermediate language for over a decade. Major architectures were supporting
uplifting to the ESIL. During the RSoC 2021 the initial version of the [new intermediate
language](https://github.com/rizinorg/rizin/blob/dev/doc/rzil.md)
that is based on the BAP's Core Theory was implemented. In the following months it was improved
and some of the architectures were ported to use RZIL instead of ESIL. The main goal of this project
is to finish the migration of the existing architectures that still use ESIL or add a RZIL support
for the architectures that hadn't any uplifting at all.

### Tasks

- Improvement a RZIL uplfiting for any non-trivial architecture, preferably that is supported by ESIL already
- Improve the integration with analysis (variables and types differences) for the chosen architecture
- Write the test cases for Rizin regression tests and improve the results.
- Implement necessary commands and APIs in Rizin for visual representation of the IL tree

> - Implement standard and graph views in Cutter for the IL output

Reviewer 2: maybe say to use the rz-tracetest from @thestr4ng3r

### Skills
Participant should know C and bits of C++ as well as be familiar with basics of the program analysis. Having an
experience with other intermediate language, SAT/SMT, and mathematical logic is a plus.

### Difficulty

> Medium

Reviewer 2: Hard/Very Hard

Reviewer 2: the hard part is to verify.

Reviewer 2: let's keep it like this and judge based on the choosen arch by the student.

Moreover, participant will become familiar with the both symbolic and concrete emulation during the
implementation process.

### Benefits for the project
Migrating most architectures will help to deprecate and remove outdated ESIL and will help improving the analysis precision.
Adding uplifting for new architectures that weren't even supported by ESIL will imrove the
analysis to even greater degree.

### Assess requirements for midterm/final evaluation
- 1st term: finish the RZIL uplifting for the chosen architecture with basic instruction-level tests
- Final term: implement all changes in the analysis code, added more complex integration tests with types analysis

### Mentors
- xvilka
- thestr4ng3r

### Links/Resources
- [RZIL-labeled issues](https://github.com/rizinorg/rizin/labels/RZIL)
- [ESIL-labeled issues](https://github.com/rizinorg/rizin/labels/RZIL)
- [ESIL to RZIL conversion tracking issue](https://github.com/rizinorg/rizin/issues/2080)
- [Cutter: IL output and graph visual representation](https://github.com/rizinorg/rizin/issues/1133)

## Debug information handling improvements (Medium project)

> Rizin already supports most of the DWARF and PDB features, including cross-platform parsing of both.
> But there are huge chunks are missing in terms of the integration an source line information.
> Current task would be to improve parsing support of both, add compatibility with `debuginfod`,
> and enhance breakpoint integration, and variable/structure printing in the debugging mode
> with the source line information.

Reviewer: ```suggestion
Rizin already supports most of the DWARF and PDB features, including cross-platform parsing of both.
However information are usually just printed to aid the reverse engineering process, but they are not 
actually used at their best. For example, you can't use them to configure a breakpoint, nor they can be
used to access variables within a function during debugging. Moreover, it is becoming more and more 
common to store DWARF information in separate files, either shipped as separate file or downloaded 
on the fly with `debuginfod`. Rizin does not support these kind of DWARF files yet.

Your task would be to improve the parsing support of both by fixing smaller bugs, add support for
separate DWARF files and `debuginfod` and enhance breakpoint integration and variable/structure
printing in debugging mode with the source information gathered from DWARF/PDB.
```

### Task
1. Support loading DWARF information from separate files and `debuginfod`

> 2. Unify source lines/types information access for DWARF, PDB, dSYM

Reviewer: ```suggestion
2. Unify source lines/types information access for DWARF, PDB, dSYM and refactor/fix parsing code as necessary
```

3. Integrate source line and types/variables information with the analysis
4. Integrate source line and types/variables with printing with `p` commands in the debug mode
5. Integrate source line and types/variables with breakpoint commands and APIs
6. Parsing performance improvements

### Skills

> Participant should know C as well as have the experience of debugging with GDB or LLDB.

Reviewer: ```suggestion
* Good knowledge of the C language
* Some experience in debugging with GDB or LLDB
* Basic knowledge of at least one of the following formats: ELF, DWARF, PDB, PE
```

### Difficulty
Hard

### Benefits for the participant
Participant will understand how high-level features of debuggers work as well as gain skills in the field of software architecture of a large, modular C project.

### Assess requirements for midterm/final evaluation
- 1st term: `debuginfod` and source line information refactoring are implemented

> - Final term: Integration of variable information with the analysis and debug printing is implemented

Reviewer: ```suggestion
- Final term: Integration of variable information with the debug and printing commands is implemented
```

### Mentors
 - xvilka
 - thestr4ng3r
 - ret2libc

### Links/Resources
- [Loading debug information from `debuginfod`](https://github.com/rizinorg/rizin/issues/1285)
- [Unify code of source information access for DWARF, PDB, dSYM](https://github.com/rizinorg/rizin/issues/907)
- [Debuginfod](https://sourceware.org/elfutils/Debuginfod)
- [Debian Debuginfod](https://wiki.debian.org/Debuginfod)
- [Fedora Debuginfod](https://fedoraproject.org/wiki/Debuginfod)
- [DWARF-labeled issues](https://github.com/rizinorg/rizin/labels/DWARF)
- [PDB-labaled issues](https://github.com/rizinorg/rizin/labels/PDB)

> ## Thread-safety and multithreading (Medium project)

Reviewer: Not a big fan of this project... I think we may have better things to propose :(

Author: Please share your ideas then.

Currently Rizin is not thread safe completely internally and as a library for multithreaded
application. The goal of this project is to eliminate global states and use contexts,
eliminate singletons, e.g. RzCons, and use thread-safe external functions and dependencies.

### Task
1. Migrate from thread-unsafe system and external dependencies
2. Eliminate global state inside RzCons and use of the singleton
3. Make RzBin thread-safe
4. Make RzAnalysis thread-safe
5. Make RzCore thread-safe
6. Add tests for using multiple RzCore and RzAnalysis instances
7. Parallelize some of the RzAnalysis function using the threading API

### Skills
Participant should know C as well as have the experience of developing multithreaded applications.

### Difficulty
Hard

### Benefits for the participant

> Participant will understand the hurdles of multithreaded programming, data syncronization, locks and debugging of such code.

Reviewer 3: A small typo here:

```suggestion
Participant will understand the hurdles of multithreaded programming, data synchronization, locks and debugging of such code.
```

### Assess requirements for midterm/final evaluation
- 1st term: `RzTypes` module exists and contains all relevant code.

> - Final term: Tree-sitter based C parser is implemented and integrated into the analysis framework.

Reviewer 2: i think this is wrong. is this a copy of last year one?

### Mentors
 - xvilka
 - thestr4ng3r
 - wargio

### Links/Resources
- [Migrate from `wcstombs()` function since it's not thread-safe](https://github.com/rizinorg/rizin/issues/1058)

## Rewriting GPL-only code (Medium project)

Currently some of the Rizin main features rely on the GPL-only code copied from binutils or GCC.
The goal is to rewrite all this code from GPL-only to LGPL or any other less restrictive license.
It is quite important for better adoption of Rizin as a library in other FOSS and commercial
projects.

## Tasks

> 1. Rewrite C++ demangler to no use GPL-only code

Reviewer 2: ```suggestion
1. Rewrite C++ demangler to and remove the GPL code
```

Reviewer 2: maybe implement the GNUv2 C++ demangler if too easy?

Author: Yes

### Skills
The participant should be comfortable with the C and Python languages, as well as have a deep understanding
of common memory management patterns such as ownership and reference counting.

### Difficulty
Advanced

### Benefits for the participant
The participant will gain an experience of exposing a C-based API with manual memory management to
high-level, object-oriented languages with automatic memory management.

### Assess requirements for midterm/final evaluation
- 1st term: Bindings can be generated as part of the standard Rizin build system and small parts of the core API are already usable.
- Final term: All relevant parts of the API can be used through bindings and also from within Cutter's Python interpreter.

### Mentors
- thestr4ng3r
- xvilka

### Links/Resources
- [SWIG Website](http://www.swig.org/)
- [SWIG 4.0 Documentation](http://www.swig.org/Doc4.0/SWIGDocumentation.html)
- [Small PoC of bindings generated in Rizin's build system](https://github.com/rizinorg/rizin/pull/285)
- [Article about Rizin's build system design](https://rizin.re/posts/why-meson/)

# Cutter

## Plugins and Python High Level API (Medium project)

> We currently don't have API almost for plugin authors to use. We need to improve a lot of things about our Plugins support and take it few steps ahead.

Reviewer 4: ```suggestion
Our current public API to be used by plugin authors is somewhat limited. We need to improve a lot of things about our Plugins support and take it few steps ahead.
```

> This task is only about improving the Python interface in Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.

Reviewer 4: ```suggestion
This task is only about improving the public C++ and Python interface of Cutter, specifically its graphical user interface components. For a task about exposing Rizin's API for disassembly, analysis and other purposes, see the Rizin bindings task above.
```

 - Python integration and IPython console.

### Skills
The participant should be comfortable with the C++ and Python languages, and be familiar with Qt framework

### Difficulty
Advanced

### Benefits for the participant
The participant will gain an experience of creating a suitable API for scripting graphical interface
programs.

### Benefits for the project
It will greatly improve the scripting experience, will make API more consistent and will ease
creating Cutter plugins by the community. Moreover, it will simplify testing of the Cutter features.

### Assess requirements for midterm/final evaluation
- 1st term: Design of the high level API and required Rizin changes. Review and implement all missing API functions that are accessible as interface controls.
- Final term: Implement the way to show the API when hovered over some interface control, create documentation.

### Mentors
- thestr4ng3r
- Megabeets

### Links/Resources
- [SDB Module/API for Cutter Python/Jupyter integration](https://github.com/rizinorg/cutter/issues/1778)
- [Jupyter plugin for Cutter](https://github.com/rizinorg/cutter-jupyter)

## Multi-Tasking and Event-driven architecture (Extended deadline)

> Cutter is a reverse engineering framework that is powered by Rizin. The information it gets about functions, strings, imports, and the analysis are all performed in Rizin and displayed in Cutter. Currently, Cutter is pulling information from Rizin only on demand. This is problematic because sometimes the user performs changes (via plugins, the console widget, and more) that are affecting the information from Rizin, but Cutter doesn't know about these changes to apply the to the UI. For example, if  a user will define a new function in a Python script or via the console widget by using the Rizin command `af @ <addr>`, Cutter will now show this new function in the Functions widget until the user will refresh the interface manually (edit -> Refresh Contents).

Reviewer 4: ```suggestion
The information Cutter gets about functions, strings, imports, and the analysis are all performed in Rizin and only displayed in Cutter. Currently, it is pulling most information from Rizin only on demand. This is problematic because sometimes the user performs changes (via plugins, the console widget, and more) that are affecting the information from Rizin, but Cutter doesn't know about these changes to apply the to the UI. For example, if a user will define a new function in a Python script or via the console widget by using the Rizin command `af @ <addr>`, Cutter will not show this new function in the Functions widget until the user will refresh the interface manually (Edit -> Refresh Contents).
The goal of this task is to use an event-driven architecture to overcome this limitation.
```

In addition, this task will also handle the analysis in the background feature, to allow the analysis performed by Rizin to happen while the interface is active.

### Tasks
The overall implementation of this task should start from Rizin by adding events to many of the functions. This can be done using `rz_events`. For example, add an even for function creating, for section creation, for flag deletion, for name changed, and more

- Add events to all the relevant functions inside Rizin
- Add support for these events in Cutter and refresh and update the relevant widgets per each event
- Support analysis in the background and allow the user to start its session while Rizin is analyzing (see [#1856](https://github.com/rizinorg/cutter/issues/1856), [#1574](https://github.com/rizinorg/cutter/issues/1574))

### Skills

> The participant should be comfortable with the C++ for Cutter and C for Rizin. The participant should be familiar with Qt framework.

Reviewer 4: ```suggestion
The participant should be comfortable with the C++ for Cutter and C for Rizin. They should also be familiar with Qt framework. Experience in GUI code architecture, for example using functional reactive programming or Elm-like approaches is a plus.
```

### Difficulty
Advanced

### Benefits for the participant
The participant will gain an experience of creating complex event-driven software in both C and C++ languages.

### Benefits for the project
It will allow to work on big files effortlessly in Cutter, will improve analysis quality as well.

### Assess requirements for midterm/final evaluation
- 1st term: Implement events everywhere in the relevant places across Rizin code and event-driven
	interaction with Cutter.
- Final term: Add support for the Cutter interface refresh based on the events from Rizin, implement analysis in background.

### Mentors
- thestr4ng3r

> - Karliss

Reviewer 4: Karliss isn't on the mentor list (I think he is not available this year, is he?)

> ## Heap viewer (Medium project)

Reviewer 4: Wasn't that partially finished last year?

Reviewer 5: There are major performance issues with the windows implementation and [Cutter integration wasn't completed](https://github.com/rizinorg/cutter/pull/2723). This could only be a medium project if support for other allocators is included.

We already have a nice heap (and memory map) parser and visualizer in Rizin (`dm` and `dmh` commands). After debugging becomes a first-class citizen in cutterland it would be awesome to have memory map and heap visualizations.

### Task
 - Expose Rizin API/commands for Cutter to use for visualization
 - Design and implement heap navigation and inspection widgets
 - Provide the integration with current debugging mode in Cutter

>  - Make the implementation work with both local (native) and remote debugging modes

Author: Yes, good idea

### Mentors
- xvilka
- Megabeets

### Links/Resources
- [Issue #1041](https://github.com/rizinorg/cutter/issues/1041)
- [Heap Viewer plugin for IDA Pro](https://github.com/danigargu/heap-viewer)
- [Heap parsing for MacOS, tmalloc, jmalloc](https://github.com/rizinorg/rizin/issues/208)
- [Dynamic Allocator Detection](https://github.com/rizinorg/rizin/issues/157)
- ["heap"-marked Rizin issues](https://github.com/rizinorg/rizin/labels/heap)

## Diffing mode (Medium project)

Binary diffing is one of the most common tasks for the reverse engineer. There are many various
tools available, but most of them are either detached from the main RE toolbox or poorly integrated.
Rizin provides basic diffing features out of the box with `rz-diff` tool, but Cutter has no
interface to represent similar functionality.

### Task
- Expose basic `rz-diff` features in the Cutter
- Create the interface to choose two files for diffing
- Create the way to show the differences in all main widgets:
	- Hexadecimal view
	- Disassembly view
	- Graph view
	- Pseudocode view

### Skills
The participant should be comfortable with the C++ language, and be familiar with Qt framework

### Difficulty

> Advanced

Reviewer 2: i think this is medium. hex diffing is slightly more difficult than asm/grap/pseudo diffing.
i would be nice to have the results of rz-diff for function similarities on 2 bins into cutter

---
title: "GSoC 2022"
layout: "single"
ShowToc: true
TocOpen: false
ShowReadingTime: false

---

**TL;DR** Jump to the [Ideas list](#project-ideas).

# Introduction

This year is the second time we participate as a fork - Rizin, effectively continuing the tradition since the year 2015 (as the radare2 project).

## Mentors

Members of the Rizin and Cutter core teams have volunteered to guide participants for GSoC’22. They were already guiding the participants for the GSoC and RSoC in past years. Please feel free to reach out to any of them in case you need any help in selecting a project.

- **Anton Kochkov** Mattermost: xvilka -- [@akochkov](https://twitter.com/akochkov)
- **Riccardo Schirone** Mattermost: ret2libc
- **Florian Märkl** Mattermost/Telegram: @thestr4ng3r -- [@thestr4ng3r](https://twitter.com/thestr4ng3r)

> - **Antide Petit** IRC/Telegram: xarkes -- [@xarkes\_](https://twitter.com/xarkes_)

Reviewer 6: Due to my current involvement in the project, I don't think I should mentor this year, so I think you can remove this line :)

> - **Antide Petit** IRC/Telegram: xarkes -- [@xarkes\_](https://twitter.com/xarkes_)

Reviewer 2: ```suggestion
```
